### PR TITLE
Add auto captcha solving

### PR DESCRIPTION
**Pull request by LucsT**
Hey nicofrand, 
We still have difficulties (captcha) in production.

We want to add auto captcha solving.
  It will work if a recaptcha key is present in the environnement (like in ours)
  If not, and a captcha is render, it will throw the same error as today ( .CAPTCHA )
  If no captcha, nothing change.

We hope that solving a captcha from time to time will help to whitelist our ip.
Moreover, one solving from time to time, allowed multiple account to fetch their bills before getting 'captched' again.

This pr add the logic and code to allowed that.
It has been tested on multiple accounts and the connector still connect and import bills without signs of regression. But please double check it with your account.

Let me know if you have questions

Lucas

### CODE REVIEW

**Comment by LucsT** [comment]
I added a second commit that change a login_failed to a vendor_down, if we can success in evaluating a login_token.
Seems preferable as is not a sign of bad credentials but more a website or connector problem

File: src/index.js [hunk 1]
```diff
@@ -16,6 +17,35 @@ const logger = {
 const baseURL = "https://secure.materiel.net";
 
 class MaterielnetKonnector extends CookieKonnector {
+    async fetch(fields) {
+        try {
+            // Try classic execution
+            await this.tryFetch(fields);
+        } catch (err) {
+            if (err.isCaptcha && err.isCaptcha === true) {
```
Comment (nicofrand) [inline src/index.js]: One of both would be enough

File: src/index.js [hunk 2]
```diff
@@ -16,6 +17,35 @@ const logger = {
...
+                const $ = cheerio.load(err.body);
+                const websiteKey = $(".g-recaptcha").data("sitekey");
+                const websiteURL = err.url;
+                // Solve captcha
+                const captchaToken = await solveCaptcha({ websiteURL, websiteKey });
+                // End login
+                await this.login(err.loginToken, fields, captchaToken);
+                try {
```
Comment (nicofrand) [inline src/index.js]: Maybe we could pass an optional parameter named `retry` default to `true` by default to `this.fetch` and reuse the function here by passing `retry` to false?

File: src/index.js [hunk 3]
```diff
@@ -63,34 +93,47 @@ class MaterielnetKonnector extends CookieKonnector {
...
-
-                if (err) {
-                    logger.debug(err.message);
-                    logger.error("Could not retrieve login token");
-                    return reject(new Error(errors.LOGIN_FAILED));
+                // Evaluate if a captcha is present
+                if (typeof(res.body) === "string" &&
+                    res.body.includes("window.renderCaptcha()")) {
+                    // We have encounter a captcha
+                    log("warn", "We detect a captcha");
```
Comment (nicofrand) [inline src/index.js]: Could we use `logger.warn` to keep it consistent?

File: src/index.js [hunk 4]
```diff
@@ -63,34 +93,47 @@ class MaterielnetKonnector extends CookieKonnector {
...
+                    return reject({
+                        "isCaptcha": true,
```
Comment (nicofrand) [inline src/index.js]: Why are the double quotes needed?

File: src/index.js [hunk 5]
```diff
@@ -63,34 +93,47 @@ class MaterielnetKonnector extends CookieKonnector {
...
+                        body: res.body,
+                        url: res.request.uri.href,
+                        loginToken: token});
```
Comment (nicofrand) [inline src/index.js]: Please put the `});` on a new line

File: src/index.js [hunk 6]
```diff
@@ -108,8 +151,11 @@ class MaterielnetKonnector extends CookieKonnector {
                             body = JSON.parse(body);
                         }
                         if (!body || !body.authenticationSuccess || !body.user) {
-                            if (typeof(body.loginForm) === "string" && body.loginForm.includes("window.renderCaptcha()")) {
+                            if (typeof(body.loginForm) === "string" &&
+                                body.loginForm.includes("window.renderCaptcha()")) {
```
Comment (nicofrand) [inline src/index.js]: Maybe we should use a constant to make it easier to modify it later if needed?

File: src/index.js (hunk 4 above)
Comment (LucsT) [reply src/index.js]: Done

File: src/index.js (hunk 3 above)
Comment (LucsT) [reply src/index.js]: Done

File: src/index.js (hunk 5 above)
Comment (LucsT) [reply src/index.js]: Done

File: src/index.js (hunk 6 above)
Comment (LucsT) [reply src/index.js]: Ok, I remove this type testing too, as it is not necessary anymore.

File: src/index.js (hunk 1 above)
Comment (LucsT) [reply src/index.js]: Yes

File: src/index.js (hunk 2 above)
Comment (LucsT) [reply src/index.js]: I'm not sure of what you mean here. What kind of retry are you talking about ? A promise retry and cactch block ?

I'm comfortable with this kind of code, but if you want we can change it.

Comment (nicofrand) [reply src/index.js]: I just thought this was doing pretty much the same thing than line 22 so we could do:

```js
    async fetch(fields, retry) {
        try {
            // Try classic execution
            await this.tryFetch(fields);
        } catch (err) {
            if (err.isCaptcha === true) {
                if (retry === false)
                    throw new Error(errors.CAPTCHA_RESOLUTION_FAILED);
                else {
                    // remaining code
```

**Comment by nicofrand** [comment]
Thanks I merged it. You removed some needed error handling parts though, I added them back, please check master again.

**Comment by LucsT** [comment]
@nicofrand 
I tested your master with sucess. You can publish it. Then i could validate it in registry.

Thank's

**Pull request merged**